src/rabbit_connection_tracking.erl

```diff
@@ -0,0 +1,340 @@
+  rabbit_log:info("Setting up a table for per-vhost connection counting on this node: ~p",
+    [tracked_connection_per_vhost_table_name_for(node())]),
+  clear_tracked_connection_tables_for_this_node(),
+  ok.
+
+
+-spec ensure_tracked_connections_table_for_this_node() -> ok.
+
+ensure_tracked_connections_table_for_this_node() ->
+  ensure_tracked_connections_table_for_node(node()).
+
+
+-spec ensure_per_vhost_tracked_connections_table_for_this_node() -> ok.
+
+ensure_per_vhost_tracked_connections_table_for_this_node() ->
+  ensure_per_vhost_tracked_connections_table_for_node(node()).
+
+
+-spec ensure_tracked_connections_table_for_node(node()) -> ok.
+
+ensure_tracked_connections_table_for_node(Node) ->
+    TableName = tracked_connection_table_name_for(Node),
+    case mnesia:create_table(TableName, [{record_name, tracked_connection},
+                                         {attributes, record_info(fields, tracked_connection)}]) of
+      {atomic, ok}     -> ok;
+      {aborted, Error} ->
+        rabbit_log:error("Failed to create a tracked connection table for node ~p: ~p", [Node, Error]),
+        ok
+    end.
+
+
+-spec ensure_per_vhost_tracked_connections_table_for_node(node()) -> ok.
+
+ensure_per_vhost_tracked_connections_table_for_node(Node) ->
+    TableName = tracked_connection_per_vhost_table_name_for(Node),
+    case mnesia:create_table(TableName, [{record_name, tracked_connection_per_vhost},
+                                          {attributes, [vhost, connection_count]}]) of
+      {atomic, ok} -> ok;
+      {aborted, _} -> ok
+      %% TODO: propagate errors
+    end.
+
+
+-spec clear_tracked_connection_tables_for_this_node() -> ok.
+
+clear_tracked_connection_tables_for_this_node() ->
+  case mnesia:clear_table(tracked_connection_table_name_for(node())) of
+      {atomic, ok} -> ok;
+      {aborted, _} -> ok
+  end,
+  case mnesia:clear_table(tracked_connection_per_vhost_table_name_for(node())) of
+      {atomic, ok} -> ok;
+      {aborted, _} -> ok
+  end.
+
+
+-spec delete_tracked_connections_table_for_node(node()) -> ok.
+
+delete_tracked_connections_table_for_node(Node) ->
+    TableName = tracked_connection_table_name_for(Node),
+    case mnesia:delete_table(TableName) of
+      {atomic, ok}              -> ok;
+      {aborted, {no_exists, _}} -> ok;
+      {aborted, Error} ->
+        rabbit_log:error("Failed to delete a tracked connection table for node ~p: ~p", [Node, Error]),
+        ok
+    end.
+
+
+-spec delete_per_vhost_tracked_connections_table_for_node(node()) -> ok.
+
+delete_per_vhost_tracked_connections_table_for_node(Node) ->
+    TableName = tracked_connection_per_vhost_table_name_for(Node),
+    case mnesia:delete_table(TableName) of
+      {atomic, ok}              -> ok;
+      {aborted, {no_exists, _}} -> ok;
+      {aborted, Error} ->
+        rabbit_log:error("Failed to delete a per-vhost tracked connection table for node ~p: ~p", [Node, Error]),
```

Comment:
We are no longer living in the age of 80 character wide IBM punch cards and terminals.